### PR TITLE
Issue 549 jarm algorithm

### PR DESCRIPTION
**Pull request by LuitelSamikshya**
```console
chaos -d hackerone.com -silent | httpx -jarm

    __    __  __       _  __
   / /_  / /_/ /_____ | |/ /
  / __ \/ __/ __/ __ \|   /
 / / / / /_/ /_/ /_/ /   |
/_/ /_/\__/\__/ .___/_/|_|
             /_/              v1.2.1-dev

		projectdiscovery.io

Use with caution. You are responsible for your actions.
Developers assume no liability and are not responsible for any misuse or damage.
https://www.hackerone.com [29d3dd00029d29d00042d43d00041d5de67cc9954cc85372523050f20b5007]
https://mta-sts.hackerone.com [29d29d00029d29d00042d43d00041d2aa5ce6a70de7ba95aef77a77b00a0af]
https://mta-sts.managed.hackerone.com [29d29d00029d29d00042d43d00041d2aa5ce6a70de7ba95aef77a77b00a0af]
https://docs.hackerone.com [29d29d00029d29d00042d43d00041d2aa5ce6a70de7ba95aef77a77b00a0af]
https://support.hackerone.com [29d3dd00029d29d00029d3dd29d29d5a74e95248e58a6162e37847a24849f7]
https://api.hackerone.com [29d3dd00029d29d00042d43d00041d5de67cc9954cc85372523050f20b5007]
http://gslink.hackerone.com []
https://mta-sts.forwarding.hackerone.com [29d29d00029d29d00042d43d00041d2aa5ce6a70de7ba95aef77a77b00a0af]
https://resources.hackerone.com [2ad2ad0002ad2ad0002ad2ad2ad2ad043bfbd87c13813505a1b60adf4f6ff5]
```

### CODE REVIEW

**Review by ehsandeep — DISMISSED** [review]
- [x] removing debug output in result
- [x] removing default value for the cases jarm hash can not be calculated.

Current:

```console
echo www.hackerone.com | ./httpx -p 80 -hash jarm -silent

{www.hackerone.com 80 0 <nil>} http://www.hackerone.com:80
http://www.hackerone.com [00000000000000000000000000000000000000000000000000000000000000]
```

Expected:

```console
echo www.hackerone.com | ./httpx -p 80 -hash jarm -silent

http://www.hackerone.com []
```

File: runner/options.go [hunk 1]
```diff
@@ -248,7 +248,7 @@ func ParseOptions() *Options {
 		flagSet.BoolVarP(&options.OutputContentType, "content-type", "ct", false, "display response content-type"),
 		flagSet.BoolVar(&options.Location, "location", false, "display response redirect location"),
 		flagSet.BoolVar(&options.Favicon, "favicon", false, "display mmh3 hash for '/favicon.ico' file"),
-		flagSet.StringVar(&options.Hashes, "hash", "", "display response body hash (supported: md5,mmh3,simhash,sha1,sha256,sha512)"),
+		flagSet.StringVar(&options.Hashes, "hash", "", "display response body hash (supported: md5,mmh3,simhash,sha1,sha256,sha512,jarm)"),
```
Comment (Mzack9999) [inline runner/options.go:251]: jarm is related to tls configuration, the description should be something like:
```
flagSet.StringVar(&options.Hashes, "hash", "", "display response body hash (supported: md5,mmh3,simhash,sha1,sha256,sha512) or tls hash (supported: jarm)"),
```

**Review by Mzack9999 — DISMISSED** [review]

File: runner/runner.go [hunk 2]
```diff
@@ -1213,7 +1213,7 @@ retry:
 	}
 	// adding default hashing for json output format
 	if r.options.JSONOutput && len(scanopts.Hashes) == 0 {
-		scanopts.Hashes = "md5,mmh3,sha256,simhash"
+		scanopts.Hashes = "md5,mmh3,sha256,simhash,jarm"
```
Comment (Mzack9999) [inline runner/runner.go:1216]: jarm shouldn't be calculated by default

File: runner/options.go (hunk 1 above)
Comment (LuitelSamikshya) [reply runner/options.go]: Done

File: runner/runner.go (hunk 2 above)
Comment (LuitelSamikshya) [reply runner/runner.go]: Done

File: common/hashes/hashes.go [hunk 3]
```diff
@@ -62,3 +65,21 @@ func Simhash(data []byte) string {
 	hash := simhash.Simhash(simhash.NewWordFeatureSet(data))
 	return fmt.Sprintf("%d", hash)
 }
+
+func Jarm(host string) string {
+	t := target{}
+	if u, err := url.Parse(host); err == nil {
```
Comment (Mzack9999) [inline common/hashes/hashes.go:71]: I think we can skip the check if the url starts with `http` and return `""` directly (no tls involved)

File: common/hashes/hashes.go [hunk 4]
```diff
@@ -62,3 +65,21 @@ func Simhash(data []byte) string {
...
+		t.Host = u.Hostname()
+		port, _ := strconv.Atoi(u.Port())
+		t.Port = port
+	}
+	if t.Port == 0 {
+		t.Port = defaultPort
+	}
+	hash := fingerprint(t)
+	rexp, _ := regexp.Compile("(?m)0{62}")
```
Comment (Mzack9999) [inline common/hashes/hashes.go:80]: as the regexp is used many times we can move its declaration + compile at local package level:

```go
package xxx

var jarmHashRegex := regexp.MustCompile("(?m)0{62}")
```

File: common/hashes/jarmhash.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+				bo = DefualtBackoff
+			}
+			time.Sleep(bo(n, t.Retries))
+			n++
+		}
+		if c == nil {
+			return ""
+		}
+		data := jarm.BuildProbe(probe)
+		_ = c.SetWriteDeadline(time.Now().Add(time.Second * 5))
```
Comment (Mzack9999) [inline common/hashes/jarmhash.go:49]: if possible we should try to respect the command line setting for connection timeout

File: common/hashes/jarmhash.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+		_, err := c.Write(data)
+		if err != nil {
+			results = append(results, "")
+			c.Close()
+			continue
+		}
+		_ = c.SetReadDeadline(time.Now().Add(time.Second * 5))
```
Comment (Mzack9999) [inline common/hashes/jarmhash.go:56]: if possible we should try to respect the command line setting for connection timeout

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

**Comment by Mzack9999** [comment]
To be confirmed:
- [ ]  moving jarm hash into an independent `tls-hash` command line parameter (cc @ehsandeep)

**Review by ehsandeep — APPROVED** [review]

**Review by Mzack9999 — APPROVED** [review]

**Pull request merged**